lib/trace/store.js

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+//
+// A lightweight store for managing tracers.
+//
+
+var EventEmitter = require('events').EventEmitter;
+
+var errors = require('../errors');
+
+var core = require('./core');
+var types = require('./types');
+
+function TracerStore(ringpop) {
+    if (!(this instanceof TracerStore)) {
+        return new TracerStore(ringpop);
+    }
+    this.ringpop = ringpop;
+    this.tracers = {};
+}
+require('util').inherits(TracerStore, EventEmitter);
+
+TracerStore.prototype.add = function add(config, opts, callback) {
+    var self = this;
+    
+    if (callback) {
+        self.once('add', callback);
+    }
+
+    var err = core.Tracer.invalidateOptions(opts);
+    if (err) {
+        self.emit('add', err);
+        return null;
+    }
+
+    var tracer = self._get(config, opts);
+    if (!tracer) {
+        var TracerType = types[opts.sink.type];
+        tracer = new TracerType(self.ringpop, config, opts);
+        var err = tracer.invalidate(opts);
+        if (err) {
+            self.emit('add', err);
+            return null;
+        }
+
+        var eventTracers = self.tracers[config.traceEvent];
+        if (!eventTracers) {
+            eventTracers = [];
+            self.tracers[config.traceEvent] = eventTracers;
+        }
+        eventTracers.push(tracer);
+
+        tracer.connectSink(function onConnect(err) {
+            if (err) {
+                self.remove(config, opts, null);
+                self.emit('add', err, null);
+                return;
+            }
+            tracer.connectSource();
+            self._updateExpiration(config, opts, tracer);
+            self.emit('add', null, tracer);
```

Comment:
I think emitting `add` like this makes for a strange API. Either you emit synchronously or async, but not both depending upon the situation. The sync emit() below deserves a nextTick. Thoughts?